File: tst/clip_pct.c
#include "pico.h"
#include "../check.h"

int main (void) {
    pico_init(1);
    pico_set_window_title("Clip");

    puts("rect at 30%");
    pico_output_clear();
    Pico_Rel_Rect r1 = { '%', {0.3,0.3, 0.5,0.5}, PICO_ANCHOR_C };
    pico_output_draw_rect(r1);
    _pico_check("clip-01");     // same as raw

    pico_set_scene_clip(r1);
    pico_set_effect_color((Pico_Color){0xCC, 0xCC, 0xCC, 0xFF});
    pico_output_clear();
    _pico_check("clip-02");     // same as raw

    puts("red centered under white");
    Pico_Rel_Rect r2 = pico_in_rect(r1, (Pico_Rel_Rect){ '%', {0.5,0.5, 0.5,0.5}, PICO_ANCHOR_C });
    pico_set_pencil_color((Pico_Color){0xFF,0x00,0x00, 0xFF});
    pico_output_draw_rect(r2);
    _pico_check("clip-03");     // same as raw

    puts("blue centered under white");
    Pico_Rel_Pos pt3 = pico_in_pos(r1, (Pico_Rel_Pos){ '%', {0.5,0.5}, PICO_ANCHOR_C });
    pico_set_pencil_color((Pico_Color){0x00,0x00,0xFF, 0xFF});
    pico_output_draw_pixel(pt3);
    _pico_check("clip-04");     // same as raw

    puts("yellow clipped under white bottom right");
    Pico_Rel_Rect r4 = pico_in_rect(r1, (Pico_Rel_Rect){ '%', {1,1, 0.5,0.5}, PICO_ANCHOR_C });
    pico_set_pencil_color((Pico_Color){0xFF,0xFF,0x00, 0xFF});
    pico_output_draw_rect(r4);
    _pico_check("clip-05");     // same as raw

    pico_init(0);
    return 0;
}

File: lua/tst/clip_pct.lua
require 'pico.check'

pico.init(true)
pico.set.window { title="Clip" }

local r1 = {'%', x=0.3, y=0.3, w=0.5, h=0.5}

do
    print("rect at 30%")
    pico.output.clear()
    pico.output.draw.rect(r1)
    pico.check("clip-01")
end

do
    print("gray background")
    pico.set.view { clip=r1 }
    pico.set.color.clear(0xCC, 0xCC, 0xCC)
    pico.output.clear()
    pico.check("clip-02")
end

do
    print("red centered under gray")
    local r2 = {'%', x=0.5, y=0.5, w=0.5, h=0.5, up=r1}
    pico.set.color.draw(0xFF, 0x00, 0x00)
    pico.output.draw.rect(r2)
    pico.check("clip-03")
end

do
    print("blue centered under gray")
    local p3 = {'%', x=0.5, y=0.5, up=r1}
    pico.set.color.draw(0x00, 0x00, 0xFF)
    pico.output.draw.pixel(p3)
    pico.check("clip-04")
end

do
    print("yellow clipped under gray bottom right")
    local r4 = {'%', x=1, y=1, w=0.5, h=0.5, up=r1}
    pico.set.color.draw(0xFF, 0xFF, 0x00)
    pico.output.draw.rect(r4)
    pico.check("clip-05")
end

pico.init(false)
